src/batch.js

```diff
@@ -2,27 +2,243 @@
+        ) {
+          arr[0].push(req)
+        } else {
+          arr[1].push(req)
+        }
+        return arr
+      },
+      [[], []]
+    )
+
+    for (const req of toExecute) {
+      idToPromise[req.id] = new Promise(resolve => {
+        ;(async () => {
+          if (req.dependsOn) {
+            for (const id of req.dependsOn) {
+              await idToPromise[id]
+            }
+          }
+
+          resolve(
+            await batch.executeRequest(route, event, context, req).catch(err => {
+              return {
+                response: {
+                  statusCode: 500,
+                  body: JSON.stringify({ statusCode: 500, message: err.message, stack: err.stack })
+                }
+              }
+            })
+          )
+        })()
+      })
+    }
+
+    //Saftey Check; Shouldn't happen, but if it were to we'd be stuck in an infinite loop
+    if (toExecute.length === 0) {
+      throw new HttpError(400, `Invalid dependency chain`)
+    }
+  } while (remainingRequests.length > 0)
+
+  const responses = []
+  const ids = Object.keys(idToPromise)
+  for (let i = 0; i < ids.length; i++) {
+    const result = await idToPromise[ids[i]]
+    responses.push({
+      id: ids[i],
+      status: result.response.statusCode,
+      body: result.response.body ? JSON.parse(result.response.body) : undefined,
+      headers: result.response.headers
+    })
+  }
+
+  return {
+    statusCode: 200,
+    body: responses,
+    headers: {}
+  }
+}
+
+function composeQueryStringParametersFromUrl(url) {
+  const urlAndQuerystring = url.split('?')
+  let queryStringParameters = {}
+  if (urlAndQuerystring.length > 1) {
+    const queryStringParams = urlAndQuerystring[1]
+    const keys = queryStringParams.split('&')
+    const queryStrings = keys.map(k => k.split('='))
+
+    for (let queryString of queryStrings) {
+      if (queryStringParameters[queryString[0]]) {
+        queryStringParameters[queryString[0]].push(queryString[1])
+      } else {
+        queryStringParameters[queryString[0]] = [queryString[1]]
+      }
+    }
+  }
+  return queryStringParameters
+}
+
+async function executeRequest(route, event, context, request) {
+  const urlAndQueryString = request.url.split('?')
+
+  let headers
+  if (event.headers) {
+    const authorizationHeader = Object.keys(event.headers).find(
+      header => header.toLowerCase() === 'authorization'
+    )
+    headers = { ...request.headers, [authorizationHeader]: event.headers[authorizationHeader] }
+  } else {
+    headers = request.headers
+  }
+
+  //Compose event to appear as native AWS event
+  let childEvent = {
+    ...event,
+    httpMethod: request.method,
+    headers,
+    body: request.body,
+    path: urlAndQueryString[0],
+    pathParameters: {
+      proxy: urlAndQueryString[0].substring(1)
+    },
+    multiValueQueryStringParameters: batch.composeQueryStringParametersFromUrl(request.url)
+  }
+
+  context._batch = true
```

Comment:
We talked about adding the `id` from the batch to the event. I'd like to avoid modifying the `context` any further

Reply:
Yes, I've swapped it for:
batchRequestId: request.id
on L#136